timesketch/lib/utils.py

```diff
@@ -101,79 +117,59 @@ def read_and_validate_csv(file_handle, delimiter=','):
-
-                normalized_timestamp = int(
-                    time.mktime(parsed_datetime.utctimetuple()) * 1000000)
-                normalized_timestamp += parsed_datetime.microsecond
-                row['timestamp'] = str(normalized_timestamp)
-                if 'tag' in row:
-                    row['tag'] = [x for x in _parse_tag_field(row['tag']) if x]
-
-                _scrub_special_tags(row)
-            except ValueError:
-                continue
-
-            yield row
-    except csv.Error as e:
+        reader = pandas.read_csv(file_handle, sep=delimiter)
+    except pandas.errors.ParserError as e:
         error_string = 'Unable to read file, with error: {0!s}'.format(e)
         logger.error(error_string)
         raise errors.DataIngestionError(error_string)
 
+    validate_csv_fields(mandatory_fields, reader)
+    for row in reader:
+        if not row:
+            continue
+        if not row['datetime']:
+            logger.warning(
+                'Row missing a datetime object, skipping [{0:s}]'.format(
+                    ','.join([str(x).replace(
+                        '\n', '').strip() for x in row.values()])))
+            continue
+        try:
+            # Normalize datetime to ISO 8601 format if it's not the case.
+            parsed_datetime = parser.parse(row['datetime'])
+            row['datetime'] = parsed_datetime.isoformat()
+
+            normalized_timestamp = int(
+                time.mktime(parsed_datetime.utctimetuple()) * 1000000)
+            normalized_timestamp += parsed_datetime.microsecond
+            row['timestamp'] = str(normalized_timestamp)
+            if 'tag' in row:
+                row['tag'] = [x for x in _parse_tag_field(row['tag']) if x]
+
+            _scrub_special_tags(row)
+        except ValueError:
+            continue
+        yield row
+
 
 def read_and_validate_redline(file_handle):
     """Generator for reading a Redline CSV file.
     Args:
         file_handle: a file-like object containing the CSV content.
+
+    Raises:
+        RuntimeError: if there are missing fields.
     """
     # Columns that must be present in the CSV file
     mandatory_fields = ['Alert', 'Tag', 'Timestamp', 'Field', 'Summary']
 
     csv.register_dialect(
-        'myDialect', delimiter=',', quoting=csv.QUOTE_ALL,
+        'redlineDialect', delimiter=',', quoting=csv.QUOTE_ALL,
         skipinitialspace=True)
-    reader = csv.DictReader(file_handle, delimiter=',', dialect='myDialect')
-
-    csv_header = reader.fieldnames
-    missing_fields = []
-    # Validate the CSV header
-    for field in mandatory_fields:
-        if field not in csv_header:
-            missing_fields.append(field)
-    if missing_fields:
-        raise RuntimeError(
-            'Missing fields in CSV header: {0:s}'.format(missing_fields))
+    reader = pandas.read_csv(file_handle, delimiter=',',
+                             dialect='redlineDialect')
```

Comment:
do we need this special dialect here? this is just indicating quoting, delimiter, etc....

Reply:
I've talked about this briefly in the PR. The issue is with the quoting parameter which either takes a csv.QUOTE or a magic number

Reply:
ok, so let's keep the special dialect in here then